static/js/place_template.tsx

```diff
@@ -121,53 +121,80 @@ class ParentPlace extends Component<ParentPlacePropsType, {}> {
 }
 
 interface RankingPropsType {
+  dcid: string;
   data: {
     label: string[];
     Population: { name: {}; label: string }[];
   };
 }
 
-// tslint:disable-next-line: max-classes-per-file
-class Ranking extends Component<RankingPropsType, {}> {
+interface RankingStateType {
+  data: {
+    label: string[];
+    Population: { name: {}; label: string }[];
+  };
+}
+
+class Ranking extends Component<RankingPropsType, RankingStateType> {
   constructor(props) {
     super(props);
+    this.state = {
+      data: this.props.data,
```

Comment:
I'm quite confused about this. Which data is this.props.data (RankingPropsType or RankingStateType)? And why is it set here when it will get filled in by the request?

Reply:
this is RankingStateType.

the sequence of action will be constructor() -> render() -> componentDidMount()

after constructor(), render() will kick in and use this.state.data in line 148, finding the length ===0 and renders empty component. Then in componentDidMount(), it fetches data and set state.data. As state is updated, render() is called again to refresh the page.

This is kinda a typical patter where data needs to be async obtained to render the content. Effectively the same as how it is before in place.ts, where we call "axios.get().then(react.Render())"

Reply:
Got it, thanks for the explanation!

Hm, why not explicitly start it off with a blank dictionary then? Unless it's to get the right type info...

Btw, then the data property in RankingPropsType is not used anymore and should be removed.

Reply:
Good point. Remove "data" in props.